### PR TITLE
Make clang-format auto-format to our required 80 char line length.

### PR DESCRIPTION
**Pull request by CodingCanuck**
This should make clang-format auto-format code to match the limit we check in https://github.com/3rdparty/dev-tools/blob/main/check_line_length.sh.

Note that this might cause several files to be reformatted on their next edit.

e.g. this change make clang-format change this:
```
  auto o = std::make_unique<Operation>(Build(operation(5, std::move(promise1))));
```

to this:
```
  auto o =
      std::make_unique<Operation>(Build(operation(5, std::move(promise1))));
```

### CODE REVIEW

File: .clang-format [hunk 1]
```diff
@@ -32,7 +32,7 @@ BinPackParameters: false
 BreakBeforeBinaryOperators: NonAssignment
 BreakBeforeTernaryOperators: true
 BreakConstructorInitializers: BeforeColon
-ColumnLimit: 0
+ColumnLimit: 80
```
Comment (benh) [inline .clang-format:35]: Oh shoot, that's our bad, we should have had a comment here!

We actually want to respect the programmers choice of newlines because `clang-format` does some really weird formatting to appease the 80 character limit. We have a `pre-commit` hook that you can install that in addition to running `clang-format` checks the line-length after the fact, see [here](https://github.com/3rdparty/dev-tools/blob/main/pre-commit#L78). There is a comment there about why we do this extra line length check, but we should have had a comment in this spot too! Would you mind adding the comment? Perhaps we just suggest reading the `dev-tools/README.md` on how to "install" the `pre-commit` hook? Happy to chat about this synchronously too if that would help.

**Review by benh — CHANGES_REQUESTED** [review]

File: .clang-format (hunk 1 above)
Comment (rjhuijsman) [reply .clang-format:35]: I'd love to re-open that discussion briefly - I at some point worked with one of our repos that had itself set to a 100-character auto-format limit, and didn't run into any case that I thought was stupid. Would love to recap the really weird formatting that you saw so that I know why I need to keep this penguin on the iceberg. 🐧 😄

Comment (CodingCanuck) [reply .clang-format:35]: Thanks Ben, I'm happy to improve comments around the current setup!

That pre-commit hook is what alerted me to this when I sent out the PR, but it would be great to figure out how I could set that up to run on local commits (so I catch these errors earlier by [shifting them left](https://www.bmc.com/blogs/what-is-shift-left-shift-left-testing-explained/)).


I'd also love to re-open this discussion :) I've found that it's quite efficient to be able to rely on the auto-formatter as much as possible: my brain power is quite limited, and it's nice to minimize brain power spent on formatting!

I'd be happy to expand this PR to include the sample changes that it would make to all files in this repo, including any hand-tweaking to keep code looking "nice" if it would be useful to have concrete diffs to talk over.

Here's an example reformatting that happens with this new clang-format setup that I personally like:

Before:
```
TEST(EventualTest, Succeed) {
  // Using mocks to ensure fail and stop callbacks don't get invoked.
  MockFunction<void()> fail, stop;

  EXPECT_CALL(fail, Call())
      .Times(0);

  EXPECT_CALL(stop, Call())
      .Times(0);

  auto e = [&]() {
    return Eventual<int>()
               .context(5)
               .start([](auto& context, auto& k) {
                 auto thread = std::thread(
                     [&context, &k]() mutable {
                       k.Start(context);
                     });
                 thread.detach();
               })
        | Then([](int i) { return i + 2; })
        | Eventual<int>()
              .context(9)
              .start([](auto& context, auto& k, auto&& value) {
                auto thread = std::thread(
                    [value, &context, &k]() mutable {
                      k.Start(context - value);
                    });
                thread.detach();
              })
              .fail([&](auto&, auto&, auto&&) {
                fail.Call();
              })
              .stop([&](auto&, auto&) {
                stop.Call();
              });
  };

  EXPECT_EQ(2, *e());
}
```

After:
```
TEST(EventualTest, Succeed) {
  // Using mocks to ensure fail and stop callbacks don't get invoked.
  MockFunction<void()> fail, stop;

  EXPECT_CALL(fail, Call()).Times(0);

  EXPECT_CALL(stop, Call()).Times(0);

  auto e = [&]() {
    return Eventual<int>().context(5).start([](auto& context, auto& k) {
      auto thread = std::thread([&context, &k]() mutable { k.Start(context); });
      thread.detach();
    }) | Then([](int i) { return i + 2; })
        | Eventual<int>()
              .context(9)
              .start([](auto& context, auto& k, auto&& value) {
                auto thread = std::thread([value, &context, &k]() mutable {
                  k.Start(context - value);
                });
                thread.detach();
              })
              .fail([&](auto&, auto&, auto&&) { fail.Call(); })
              .stop([&](auto&, auto&) { stop.Call(); });
  };

  EXPECT_EQ(2, *e());
}
```

Here's another diff that I like less:

Before:
```
TEST(EventualTest, Catch) {
  auto e = []() {
    return Just(41)
        | Raise("error")
        | Catch([](auto&& error) {
             return Just(42);
           });
  };

  EXPECT_EQ(42, *e());
}
```

After:
```
TEST(EventualTest, Catch) {
  auto e = []() {
    return Just(41) | Raise("error")
        | Catch([](auto&& error) { return Just(42); });
  };

  EXPECT_EQ(42, *e());
}
```

Having both Just() and Raise() on one line whereas Catch() is on its own line does break the flow a bit.

Here's how you could achieve manual formatting while still letting clang-format control line lengths by using comments:
```
TEST(EventualTest, Catch) {
  auto e = []() {
    return Just(41) //
        | Raise("error") //
        | Catch([](auto&& error) { return Just(42); });
  };

  EXPECT_EQ(42, *e());
}
```

You can also disable clang-format for chunks where you find it disagreeable, while still delegating formatting to it (freeing up some brainpower) in the common case:
```
TEST(EventualTest, Catch) {
  auto e = []() {
    // clang-format off
    return Just(41)
        | Raise("error")
        | Catch([](auto&& error) {
             return Just(42);
           });
    // clang-format on
  };

  EXPECT_EQ(42, *e());
}
```

(If we do want to proceed with a change like this, my current guess is that it would minimize friction to first submit changes to each relevant repo to reformat them with the updated clang-format config before actually committing the config change, then submit the config change afterwards once all existing code is compliant)

Comment (rjhuijsman) [reply .clang-format:35]: Ah, if I recall correctly, the reformatting to...
```
    return Just(41) | Raise("error")
        | Catch([](auto&& error) { return Just(42); });
```
... is core to what Ben argues is Not Nice here, despite his normal stance of very heavily favoring auto-formatting tooling.

If we see our `eventuals` repo(s) as not just working for ourselves but also as code samples for everyone else discovering this new way of working, would we accept `//\n` and `// clang-format off` in many places?

I'm a "maybe or maybe not" on auto-formatting-line-length IF code sample-ness is truly an issue. For e.g. the Respect library I care much more about our own ease of building than being a code sample for `eventuals`, so I would definitely prefer auto-line-formatting there.

Comment (benh) [reply .clang-format:35]: Yes, the reformatting highlighted by @rjhuijsman and that @CodingCanuck called out as "breaking the flow" is an example of where it goes poorly.

And yes, we could let folks turn off formatting with `// clang-format off` blocks for an example such as this, but I'm worried that this opens up new challenges.

First, we'll need to document when it's allowed to use the `// clang-format off` block, i.e., "only for eventuals code that is harder to read because it breaks the flow". Contributors will either have to read those rules ahead of time or they will have to get taught that when a reviewer suggests some line breaks (aka, a 🐧). This is not what most reviewers want to spend time on, and more importantly it puts a human in the line of fire rather than a tool (and if we could build a tool we wouldn't be having this discussion).

Second, whoever is reviewing the code inside `// clang-format off` blocks will need to look extra carefully that no other code formatting violations are  made. Honestly, there are a lot of parts of the `.clang-format` file that I don't have memorized (aka, 🐧, 🐧, 🐧) and trying to "be" `clang-format` inside those blocks does not sound like a good use of my time. As pointed out, we could use trailing empty comments `//` to ensure the line breaks while getting the rest of the code formatted correctly, but that "trick" will also have to be learned from reading documentation or taught during a review, aka, a  🐧.

Is having a human "teach" another programmer how to keep the formatter from "getting in their way" a better tradeoff than a tool "teaching" a programmer there is a character limit?

While our experience with eventuals is still young, I don't think this has been an issue thus far at all. _On the flip side_, the natural tendency that I've observed is that programmers _prefer_ to put each composed eventual (i.e., `| SomeEventual()`) on a newline, and that's not something we've had to teach! And even if we did decide to "teach it", it would be consistent with other [style guides](https://google.github.io/styleguide/shellguide.html#pipelines) that have similar functionality.

In a world where the programmer needs to keep the formatter from "getting in their way", I expect there to be _a lot_ of trailing empty comments, and even expect some programmers to do it as a _default_, because they find one eventual per line to just be more "natural". As reviewers, would we fault them?

Bottom line, I think we need to get more experience before making any changes here. If it turns out that this ends up truly being a huge time waste then we revisit. In the mean time, for anyone such as @CodingCanuck that ventures into the `.clang-format` file confused as to why `clang-format` didn't wrap their line at the column limit, I think we should add a comment that (1) explains the choice (even pointing to this discussion) and (2) links to our `pre-commit` hook so they can get notified before creating a PR.

Comment (benh) [reply .clang-format:35]: For those of you following along wondering what the heck are 🐧, they are penguins, in reference to this [article](https://jqpublicblog.com/penguins-and-icebergs/).
